Backpack/src/main/java/net/skyscanner/backpack/util/TextLayoutCompat.kt

```diff
@@ -0,0 +1,182 @@
+/**
+ * Backpack for Android - Skyscanner's Design System
+ *
+ * Copyright 2018-2020 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.util
+
+import android.os.Build
+import android.text.Layout
+import android.text.StaticLayout
+import android.text.TextDirectionHeuristic
+import android.text.TextDirectionHeuristics
+import android.text.TextPaint
+import android.text.TextUtils
+import android.util.Log
+import androidx.annotation.RequiresApi
+import java.lang.reflect.Constructor
+import net.skyscanner.backpack.BuildConfig
+
+internal object TextLayoutCompat {
+
+  private const val TAG = "TextLayoutCompat"
+
+  private val creator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+    StaticLayoutCreatorApi23
+  } else {
+    StaticLayoutCreatorApi21
+  }
+
+  fun staticLayout(
+    source: CharSequence,
+    bufStart: Int = 0,
+    bufEnd: Int = source.length,
+    paint: TextPaint,
+    outerWidth: Int,
+    align: Layout.Alignment,
+    textDir: TextDirectionHeuristic = TextDirectionHeuristics.FIRSTSTRONG_LTR,
+    spacingMult: Float = 1f,
+    spacingAdd: Float = 0f,
+    includePad: Boolean,
+    ellipsize: TextUtils.TruncateAt? = null,
+    ellipsizedWidth: Int = outerWidth,
+    maxLines: Int = Int.MAX_VALUE,
+  ): StaticLayout = creator.create(
+    source = source,
+    bufStart = bufStart,
+    bufEnd = bufEnd,
+    paint = paint,
+    outerWidth = outerWidth,
+    align = align,
+    textDir = textDir,
+    spacingMult = spacingMult,
+    spacingAdd = spacingAdd,
+    includePad = includePad,
+    ellipsize = ellipsize,
+    ellipsizedWidth = ellipsizedWidth,
+    maxLines = maxLines,
+  )
+
+  interface StaticLayoutCreator {
+
+    fun create(
+      source: CharSequence,
+      bufStart: Int,
+      bufEnd: Int,
+      paint: TextPaint,
+      outerWidth: Int,
+      align: Layout.Alignment,
+      textDir: TextDirectionHeuristic,
+      spacingMult: Float,
+      spacingAdd: Float,
+      includePad: Boolean,
+      ellipsize: TextUtils.TruncateAt?,
+      ellipsizedWidth: Int,
+      maxLines: Int
+    ): StaticLayout
+  }
+
+  object StaticLayoutCreatorApi21 : StaticLayoutCreator {
+
+    private val ConstructorReference: Constructor<StaticLayout>? = try {
```

Comment:
There's a way to create multi-line StaticLayout in Android <= 23, but the API is not public. It's used, however, by all default Android components.